### PR TITLE
config/types/etcd: add warnings when some fields don't start with schema

### PR DESCRIPTION
**Pull request by dgonyeo**
Some fields in etcd must include schemas in the given URL (http, https,
unix, or unixs), this commit adds a validation field on the options
structs that will exmit warnings if these fields don't appear to start
with the correct values.

Related to https://github.com/coreos/bugs/issues/1952#issuecomment-299489651

### CODE REVIEW

File: config/types/etcd.go [hunk 1]
```diff
@@ -142,6 +145,39 @@ func etcdContents(etcd Etcd, platform string) (string, error) {
 	return assembleUnit("/usr/lib/coreos/etcd-wrapper $ETCD_OPTS", args, vars, platform)
 }
 
+var validPrefixes = []string{"http://", "https://", "unix://", "unixs://"}
+
+func checkForPrefix(s string) bool {
+	hasPrefix := false
+	for _, p := range validPrefixes {
+		hasPrefix = hasPrefix || strings.HasPrefix(s, p)
```
Comment (crawford) [inline config/types/etcd.go]: Use the URL package instead.

File: config/types/etcd.go [hunk 2]
```diff
@@ -32,7 +33,9 @@ var (
 )
 
 // Options can be the options for any Etcd version
-type Options interface{}
+type Options interface {
```
Comment (crawford) [inline config/types/etcd.go]: Why is this being added?

File: config/types/flannel.go [hunk 3]
```diff
@@ -113,6 +113,10 @@ func flannelContents(flannel Flannel, platform string) (string, error) {
 	return assembleUnit("/usr/lib/coreos/flannel-wrapper $FLANNEL_OPTS", args, vars, platform)
 }
 
+func (f Flannel0_7) Validate() report.Report {
```
Comment (crawford) [inline config/types/flannel.go]: Why are these being added?

File: config/types/etcd.go [hunk 4]
```diff
@@ -142,6 +145,39 @@ func etcdContents(etcd Etcd, platform string) (string, error) {
...
+
+func (e Etcd3_0) Validate() report.Report {
+	r := report.Report{}
+	addPrefixWarning := func(name string) {
+		r.Add(report.Entry{
+			Message: fmt.Sprintf("%s doesn't start with one of: %v", name, validPrefixes),
+			Kind:    report.EntryWarning,
+		})
+	}
+	if !checkForPrefix(e.ListenPeerUrls) {
```
Comment (crawford) [inline config/types/etcd.go]:
```go
for key, attr in map[string]string {
    "listen_peer_urls":            e.ListenPeerUrls,
    "listen_client_urls":          e.ListenClientUrls,
    "initial_advertise_peer_urls": e.InitialAdvertisePeerUrls,
    "advertise_client_urls":       e.AdvertiseClientUrls,
} {
    if !checkForPrefix(attr) {
        r.Add(report.Entry{
            Message: fmt.Sprintf("value for %s doesn't start with one of: %v", prefix, validPrefixes),
            Kind:    report.EntryWarning,
        })
    }
}
```

You probably want to put that in its own function to avoid duplication between the different Validate methods.

File: config/types/etcd.go (hunk 2 above)
Comment (dgonyeo) [inline config/types/etcd.go]: This is necessary for the reflection logic to properly check this field for a `Validate` function. Without this the added validation logic to check for the schema prefixes is never called.

File: config/types/flannel.go (hunk 3 above)
Comment (dgonyeo) [inline config/types/flannel.go]: These are necessary for them to still implement the `Options` interface

File: config/types/etcd.go (hunk 1 above)
Comment (dgonyeo) [inline config/types/etcd.go]: The URL package chokes on dynamic data:
```
parse http://{PRIVATE_IPV4}:2380: invalid character "{" in host name
```

**Comment by dgonyeo** [comment]
Refactored a little, got tests passing, and got it to handle comma separated lists of URLs

File: config/types/etcd.go [hunk 5]
```diff
@@ -32,7 +33,9 @@ var (
 )
 
 // Options can be the options for any Etcd version
```
Comment (crawford) [inline config/types/etcd.go]: Not part of this PR, but "etcd" is always lowercase.

Comment (dgonyeo) [inline config/types/etcd.go]: I blame @ajeddeloh for that

File: config/types/etcd.go (hunk 2 above)
Comment (crawford) [inline config/types/etcd.go]: That doesn't seem right. The validation code already has this interface declared and uses that to do type-casts.

File: config/types/etcd.go [hunk 6]
```diff
@@ -147,6 +150,47 @@ func etcdContents(etcd Etcd, platform string) (string, error) {
...
+	}
+	return true
+}
+
+func validatePrefixes(m map[string]string) report.Report {
+	r := report.Report{}
+	for name, val := range m {
+		if !checkForPrefix(val) {
+			r.Add(report.Entry{
+				Message: fmt.Sprintf("%s doesn't start with one of: %v", name, validPrefixes),
```
Comment (crawford) [inline config/types/etcd.go]: The field name will never start with one of those prefixes. "value for %s doesn't..."

File: config/types/etcd.go [hunk 7]
```diff
@@ -147,6 +150,47 @@ func etcdContents(etcd Etcd, platform string) (string, error) {
...
+
+func checkForPrefix(s string) bool {
+	if s == "" {
+		return true
+	}
+	urlStrs := strings.Split(s, ",")
+	for _, urlStr := range urlStrs {
+		hasPrefix := false
+		for _, p := range validPrefixes {
+			hasPrefix = hasPrefix || strings.HasPrefix(urlStr, p)
```
Comment (crawford) [inline config/types/etcd.go]: I asked before. Use the URL package.

Comment (dgonyeo) [inline config/types/etcd.go]: The URL package chokes on dynamic data:
```
parse http://{PRIVATE_IPV4}:2380: invalid character "{" in host name
```

File: config/types/etcd.go (hunk 2 above)
Comment (dgonyeo) [inline config/types/etcd.go]: I didn't dig into the implementation of the validation logic, but the code wasn't called without this change. I can spend some time digging into the validation logic if you'd prefer me to try to fix it in there, but I have a feeling it's due to it not being able to know what functions are available to call on an empty interface.

Comment (crawford) [inline config/types/etcd.go]: But that's how Go works. All of this is determined at runtime instead of at compile time. You don't specify which interfaces you are implementing (e.g. Rust's `impl`), it uses duck-typing to figure out which interfaces are implicitly satisfied. Maybe that is broken across packages.

File: config/types/etcd.go (hunk 7 above)
Comment (crawford) [inline config/types/etcd.go]: I'm not familiar with the full URL RFC. You'll need to review that before this is safe to do. Otherwise, substitute a dummy IP address before validation.

File: config/types/etcd.go (hunk 5 above)
Comment (ajeddeloh) [reply config/types/etcd.go]: whoops.

File: config/types/etcd.go (hunk 2 above)
Comment (ajeddeloh) [reply config/types/etcd.go]: This is hitting a weird edge case with the validation code. The validation code "flattens" all embedded structs, so its checking if you can call Validate() on the top level `Etcd` struct. Go's struct embeddeding rules allows that if an embedded member allows that, but only superficially (doesn't work with []T or map[T1]T2). Unfortunately (for us), go uses the interface supplied not the underlying data's types. 

Why not use the Validate<NAME>() style functions? The flattening *should* handle that just fine.

**Comment by dgonyeo** [comment]
I'm leaving Red Hat, and thus won't be getting this merged in. If someone else wants to open a new PR with these changes, feel free.
